{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = '' # Use the data folder provided in Tutorial 02 - Intro to Pandas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "\n",
    "DATA_FOLDER = 'Data/ebola/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Guinea\n",
    "\n",
    "### Check structures of the files\n",
    "\n",
    "#### Headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>0</th>\n",
       "      <th>0</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Date</td>\n",
       "      <td>Date</td>\n",
       "      <td>Date</td>\n",
       "      <td>Date</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Description</td>\n",
       "      <td>Description</td>\n",
       "      <td>Description</td>\n",
       "      <td>Description</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Totals</td>\n",
       "      <td>Totals</td>\n",
       "      <td>Totals</td>\n",
       "      <td>Totals</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Conakry</td>\n",
       "      <td>Conakry</td>\n",
       "      <td>Conakry</td>\n",
       "      <td>Conakry</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gueckedou</td>\n",
       "      <td>Gueckedou</td>\n",
       "      <td>Gueckedou</td>\n",
       "      <td>Gueckedou</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Macenta</td>\n",
       "      <td>Macenta</td>\n",
       "      <td>Macenta</td>\n",
       "      <td>Macenta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Dabola</td>\n",
       "      <td>Dabola</td>\n",
       "      <td>Dabola</td>\n",
       "      <td>Dabola</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Kissidougou</td>\n",
       "      <td>Kissidougou</td>\n",
       "      <td>Kissidougou</td>\n",
       "      <td>Kissidougou</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Dinguiraye</td>\n",
       "      <td>Dinguiraye</td>\n",
       "      <td>Dinguiraye</td>\n",
       "      <td>Dinguiraye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Telimele</td>\n",
       "      <td>Telimele</td>\n",
       "      <td>Telimele</td>\n",
       "      <td>Telimele</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Boffa</td>\n",
       "      <td>Boffa</td>\n",
       "      <td>Boffa</td>\n",
       "      <td>Boffa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Kouroussa</td>\n",
       "      <td>Kouroussa</td>\n",
       "      <td>Kouroussa</td>\n",
       "      <td>Kouroussa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Dubreka</td>\n",
       "      <td>Siguiri</td>\n",
       "      <td>Siguiri</td>\n",
       "      <td>Siguiri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Siguiri</td>\n",
       "      <td>Pita</td>\n",
       "      <td>Pita</td>\n",
       "      <td>Pita</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Pita</td>\n",
       "      <td>Mzerekore</td>\n",
       "      <td>Nzerekore</td>\n",
       "      <td>Nzerekore</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Nzerekore</td>\n",
       "      <td>Yomou</td>\n",
       "      <td>Yomou</td>\n",
       "      <td>Yomou</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Dubreka</td>\n",
       "      <td>Dubreka</td>\n",
       "      <td>Dubreka</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Forecariah</td>\n",
       "      <td>Forecariah</td>\n",
       "      <td>Forecariah</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Kerouane</td>\n",
       "      <td>Kerouane</td>\n",
       "      <td>Kerouane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Coyah</td>\n",
       "      <td>Coyah</td>\n",
       "      <td>Coyah</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Dalaba</td>\n",
       "      <td>Dalaba</td>\n",
       "      <td>Dalaba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Beyla</td>\n",
       "      <td>Beyla</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Kindia</td>\n",
       "      <td>Kindia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Lola</td>\n",
       "      <td>Lola</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0            0            0            0\n",
       "0          Date         Date         Date         Date\n",
       "1   Description  Description  Description  Description\n",
       "2        Totals       Totals       Totals       Totals\n",
       "3       Conakry      Conakry      Conakry      Conakry\n",
       "4     Gueckedou    Gueckedou    Gueckedou    Gueckedou\n",
       "5       Macenta      Macenta      Macenta      Macenta\n",
       "6        Dabola       Dabola       Dabola       Dabola\n",
       "7   Kissidougou  Kissidougou  Kissidougou  Kissidougou\n",
       "8    Dinguiraye   Dinguiraye   Dinguiraye   Dinguiraye\n",
       "9      Telimele     Telimele     Telimele     Telimele\n",
       "10        Boffa        Boffa        Boffa        Boffa\n",
       "11    Kouroussa    Kouroussa    Kouroussa    Kouroussa\n",
       "12      Dubreka      Siguiri      Siguiri      Siguiri\n",
       "13      Siguiri         Pita         Pita         Pita\n",
       "14         Pita    Mzerekore    Nzerekore    Nzerekore\n",
       "15    Nzerekore        Yomou        Yomou        Yomou\n",
       "16          NaN      Dubreka      Dubreka      Dubreka\n",
       "17          NaN   Forecariah   Forecariah   Forecariah\n",
       "18          NaN     Kerouane     Kerouane     Kerouane\n",
       "19          NaN        Coyah        Coyah        Coyah\n",
       "20          NaN       Dalaba       Dalaba       Dalaba\n",
       "21          NaN        Beyla        Beyla          NaN\n",
       "22          NaN       Kindia       Kindia          NaN\n",
       "23          NaN         Lola         Lola          NaN"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allFiles = glob.glob(DATA_FOLDER + 'guinea_data/*.csv')\n",
    "\n",
    "def check_headers(files):\n",
    "    list_ = []\n",
    "    for file_ in files:\n",
    "        df = pd.read_csv(file_, header=None)\n",
    "        list_.append(df.head(n=1))\n",
    "\n",
    "    frame = pd.concat(list_)\n",
    "    # Transpose to not drop record and better visualisation\n",
    "    return frame.drop_duplicates().T\n",
    "\n",
    "check_headers(allFiles)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rows title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sizes: ['42', '33']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Total number of admissions to CTE',\n",
       " nan,\n",
       " 'Number of contacts out of the track 21 days',\n",
       " 'Number of samples collected today',\n",
       " 'New contacts registered so far',\n",
       " 'Total samples tested',\n",
       " 'Total suspected non-class cases',\n",
       " 'Fatality rate for confirmed and probables',\n",
       " 'Number of samples under test',\n",
       " 'Total number of exits from CTE',\n",
       " 'New cases of confirmed among health workers',\n",
       " 'Total case of confirmed among health workers',\n",
       " 'New deaths registered',\n",
       " 'Number of contacts followed today',\n",
       " 'Total number of hospitalized cases in CTE',\n",
       " 'New admits to CTE so far',\n",
       " 'New deaths registered among health workers',\n",
       " 'Number of contacts lost to follow up',\n",
       " 'Total deaths registered among health workers'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Footers, check if the description are the same for each file\n",
    "list_ = []\n",
    "for file_ in allFiles:\n",
    "    df = pd.read_csv(file_, usecols=['Description'])\n",
    "    #print(file_, df.shape)\n",
    "    list_.append(df)\n",
    "    \n",
    "frame = pd.concat(list_, axis=1)\n",
    "cleaned = frame.T.drop_duplicates().T\n",
    "\n",
    "first_column = set(cleaned[[0]].squeeze().tolist())\n",
    "second_column = set(cleaned[[1]].squeeze().tolist())\n",
    "print(\"sizes: {}\".format(list(map(str, map(len, [first_column, second_column])))))\n",
    "# Extra columns\n",
    "second_column - first_column\n",
    "\n",
    "# Let's discover which files doesn't follow the pattern\n",
    "# Check if data is already the same in each file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Headers:** Therefore, sometimes, region are missing, but the first three columns are always available: Date, Description and Totals. If we see some disrepancies between totals and the sum of all remaining columns, we are going to collect them.\n",
    "\n",
    "**Rows:** \"New deaths registered today\" and \"New deaths registered\" will be considerer as the same data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Check date per file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def are_date_unique_per_file():\n",
    "    for file_ in allFiles:\n",
    "        df = pd.read_csv(file_, usecols=['Date'])\n",
    "        #print(file_, df.shape)\n",
    "        if df.drop_duplicates().shape[0] > 1:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "are_date_unique_per_file()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the data\n",
    "\n",
    "Per file, we are going to compare the Totals colun and all region (if they sum up correctly) for the row: \"New deaths registered today (confirmed)\" or \"New deaths registered\" per file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           Totals\n",
      "Date             \n",
      "2014-08-26      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-08-27      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-08-30      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-08-31      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-02      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-04      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-07      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-08      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-09      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-11      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-14      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-16      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-17      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-19      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-21      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-22      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-23      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-24      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-26      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-09-30      0\n",
      "           Totals\n",
      "Date             \n",
      "2014-10-01      0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Deaths (via total)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-08-04</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-26</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-27</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-30</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-31</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-02</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-04</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-07</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-08</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-09</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-11</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-14</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-16</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-17</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-19</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-21</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-22</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-23</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-24</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-26</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-30</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-10-01</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Deaths (via total)\n",
       "Date                         \n",
       "2014-08-04                  2\n",
       "2014-08-26                  5\n",
       "2014-08-27                  2\n",
       "2014-08-30                  5\n",
       "2014-08-31                  3\n",
       "2014-09-02                  5\n",
       "2014-09-04                  5\n",
       "2014-09-07                  4\n",
       "2014-09-08                  4\n",
       "2014-09-09                  7\n",
       "2014-09-11                  2\n",
       "2014-09-14                  1\n",
       "2014-09-16                  3\n",
       "2014-09-17                  3\n",
       "2014-09-19                  5\n",
       "2014-09-21                  0\n",
       "2014-09-22                  3\n",
       "2014-09-23                  3\n",
       "2014-09-24                  3\n",
       "2014-09-26                  5\n",
       "2014-09-30                  4\n",
       "2014-10-01                 15"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write your answer here\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "DATA_FOLDER = 'Data/ebola/'\n",
    "\n",
    "#Guinea\n",
    "allFiles = glob.glob(DATA_FOLDER + 'guinea_data/*.csv')\n",
    "frame = pd.DataFrame()\n",
    "\n",
    "list_ = []\n",
    "for file_ in allFiles:\n",
    "    df = pd.read_csv(file_, index_col=['Description', 'Date'], usecols=['Date', 'Description', 'Totals'])\n",
    "    if \"New deaths registered\" in df.index:\n",
    "        print(df.loc[\"New deaths registered among health workers\"])\n",
    "\n",
    "        df = df.loc[\"New deaths registered\"]\n",
    "    else:\n",
    "        df = df.loc[\"New deaths registered today\"]\n",
    "    \n",
    "    list_.append(df)\n",
    "    \n",
    "# New deaths registered among health workers always zero!\n",
    "    \n",
    "    \n",
    "\n",
    "#df[[\"New deaths registered today (confirmed)\"]]\n",
    "#df.loc[df.Description == \"New deaths registered today\"]\n",
    "frame = pd.concat(list_)\n",
    "frame.columns = [\"Deaths (via total)\"]\n",
    "frame \n",
    "\n",
    "#with pd.option_context('display.max_rows', None, 'display.max_columns', 3):\n",
    "#frame.sort_index()\n",
    "#frame.index[0]\n",
    "#rame.index = frame.index.map(lambda x: (\"-\".join(x.split('-')[0:2])))\n",
    "#rame.astype(int).groupby(['Date']).sum()/30\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Death</th>\n",
       "      <th>Death (manual sum)</th>\n",
       "      <th>New cases</th>\n",
       "      <th>New cases (manual sum)</th>\n",
       "      <th>Country</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Date</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2014-08-04</th>\n",
       "      <td>0.064516</td>\n",
       "      <td>0.064516</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-26</th>\n",
       "      <td>0.161290</td>\n",
       "      <td>0.161290</td>\n",
       "      <td>0.322581</td>\n",
       "      <td>0.322581</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-27</th>\n",
       "      <td>0.064516</td>\n",
       "      <td>0.064516</td>\n",
       "      <td>0.322581</td>\n",
       "      <td>0.322581</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-30</th>\n",
       "      <td>0.161290</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>0.290323</td>\n",
       "      <td>0.290323</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-08-31</th>\n",
       "      <td>0.096774</td>\n",
       "      <td>0.096774</td>\n",
       "      <td>0.935484</td>\n",
       "      <td>0.935484</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-02</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-04</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.566667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-07</th>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-08</th>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-09</th>\n",
       "      <td>0.233333</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-11</th>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.066667</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-14</th>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-16</th>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-17</th>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-19</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-21</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.566667</td>\n",
       "      <td>0.566667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-22</th>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-23</th>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-24</th>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-26</th>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-09-30</th>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2014-10-01</th>\n",
       "      <td>0.483871</td>\n",
       "      <td>0.483871</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>Guinea</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Death  Death (manual sum)  New cases  New cases (manual sum)  \\\n",
       "Date                                                                          \n",
       "2014-08-04  0.064516            0.064516   0.129032                0.129032   \n",
       "2014-08-26  0.161290            0.161290   0.322581                0.322581   \n",
       "2014-08-27  0.064516            0.064516   0.322581                0.322581   \n",
       "2014-08-30  0.161290            0.129032   0.290323                0.290323   \n",
       "2014-08-31  0.096774            0.096774   0.935484                0.935484   \n",
       "2014-09-02  0.166667            0.166667   0.466667                0.333333   \n",
       "2014-09-04  0.166667            0.166667   0.566667                0.566667   \n",
       "2014-09-07  0.133333            0.133333   0.366667                0.266667   \n",
       "2014-09-08  0.133333            0.133333   0.366667                0.366667   \n",
       "2014-09-09  0.233333            0.233333   0.233333                0.233333   \n",
       "2014-09-11  0.066667            0.066667   0.433333                0.433333   \n",
       "2014-09-14  0.033333            0.033333   0.466667                0.466667   \n",
       "2014-09-16  0.100000            0.100000   0.200000                0.200000   \n",
       "2014-09-17  0.100000            0.100000   0.200000                0.200000   \n",
       "2014-09-19  0.166667            0.166667   0.400000                0.400000   \n",
       "2014-09-21  0.000000            0.000000   0.566667                0.566667   \n",
       "2014-09-22  0.100000            0.100000   0.466667                0.466667   \n",
       "2014-09-23  0.100000            0.100000   0.833333                0.833333   \n",
       "2014-09-24  0.100000            0.100000   0.666667                0.666667   \n",
       "2014-09-26  0.166667            0.166667   0.333333                0.333333   \n",
       "2014-09-30  0.133333            0.133333   0.366667                0.366667   \n",
       "2014-10-01  0.483871            0.483871   0.193548                0.193548   \n",
       "\n",
       "           Country  \n",
       "Date                \n",
       "2014-08-04  Guinea  \n",
       "2014-08-26  Guinea  \n",
       "2014-08-27  Guinea  \n",
       "2014-08-30  Guinea  \n",
       "2014-08-31  Guinea  \n",
       "2014-09-02  Guinea  \n",
       "2014-09-04  Guinea  \n",
       "2014-09-07  Guinea  \n",
       "2014-09-08  Guinea  \n",
       "2014-09-09  Guinea  \n",
       "2014-09-11  Guinea  \n",
       "2014-09-14  Guinea  \n",
       "2014-09-16  Guinea  \n",
       "2014-09-17  Guinea  \n",
       "2014-09-19  Guinea  \n",
       "2014-09-21  Guinea  \n",
       "2014-09-22  Guinea  \n",
       "2014-09-23  Guinea  \n",
       "2014-09-24  Guinea  \n",
       "2014-09-26  Guinea  \n",
       "2014-09-30  Guinea  \n",
       "2014-10-01  Guinea  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_ = []\n",
    "allFiles = glob.glob(DATA_FOLDER + 'guinea_data/*.csv')\n",
    "#allFiles = [\"Data/ebola/guinea_data/2014-09-22.csv\"]\n",
    "\n",
    "   \n",
    "# column titles    \n",
    "DEATHS = \"Death\"\n",
    "DEATHS_MANUAL_SUM = \"Death (manual sum)\"\n",
    "NEW_CASES = \"New cases\"\n",
    "NEW_CASES_MANUAL_SUM = \"New cases (manual sum)\"\n",
    "divisible_fields = [DEATHS, DEATHS_MANUAL_SUM, NEW_CASES, NEW_CASES_MANUAL_SUM]\n",
    "\n",
    "for file_ in allFiles:\n",
    "    df = pd.read_csv(file_, index_col=['Description', 'Date']).fillna(value=0)\n",
    "    \n",
    "    #print(df.dtypes)\n",
    "    \n",
    "    \n",
    "    if \"New deaths registered\" in df.index:\n",
    "        deaths = df.loc[\"New deaths registered\"].copy()\n",
    "    else:\n",
    "        deaths = df.loc[\"New deaths registered today\"].copy()\n",
    "        \n",
    "    new_cases = df.loc[\"New cases of confirmed\"].copy()\n",
    "    \n",
    "    #print(new_cases)\n",
    "\n",
    "    deaths_per_region = pd.to_numeric(deaths.drop(\"Totals\", axis=1).values[0]).sum().astype(int)\n",
    "    new_cases_per_region = pd.to_numeric(new_cases.drop(\"Totals\", axis=1).values[0]).sum().astype(int)\n",
    " \n",
    "    \n",
    "    deaths[DEATHS_MANUAL_SUM] = deaths_per_region\n",
    "    deaths[NEW_CASES] = int(new_cases[\"Totals\"][0])\n",
    "    deaths[NEW_CASES_MANUAL_SUM] = new_cases_per_region\n",
    "\n",
    "\n",
    "    \n",
    "    #deaths = deaths[[\"Totals\", \"Death\"]]\n",
    "    #deaths.columns = [\"Deaths\", \"Deaths (manual sum)\"]\n",
    "    \n",
    "    deaths = deaths[[\"Totals\", DEATHS_MANUAL_SUM, NEW_CASES, NEW_CASES_MANUAL_SUM]]\n",
    "    deaths.columns= [DEATHS, DEATHS_MANUAL_SUM, NEW_CASES, NEW_CASES_MANUAL_SUM]\n",
    "    \n",
    "    deaths[DEATHS] = deaths[DEATHS].astype(int)\n",
    "    \n",
    "    #print(deaths)\n",
    "    \n",
    "    list_.append(deaths)\n",
    "    \n",
    "    \n",
    "frame = pd.concat(list_)\n",
    "frame[\"Country\"] = \"Guinea\"\n",
    "frame[\"Date\"] = pd.to_datetime(frame.index, format=\"%Y-%m-%d\")\n",
    "frame\n",
    "frame.reset_index(drop=True)\n",
    "frame.set_index([\"Country\", \"Date\"])\n",
    "\n",
    "frame[\"Month+Year\"] = frame[\"Date\"].apply(lambda x: x.replace(day=1))\n",
    "frame[\"daysInMonth\"] = frame[\"Month+Year\"].apply(lambda x: int(pd.tslib.monthrange(x.year, x.month)[1]))\n",
    "\n",
    "\n",
    "#frame.groupby(\"Month+Year\").sum() / frame[\"daysInMonth\"]\n",
    "frame[divisible_fields] = frame[divisible_fields].div(frame[\"daysInMonth\"].values,axis=0)\n",
    "frame = frame.drop([\"Month+Year\", \"daysInMonth\", \"Date\"], axis=1)\n",
    "#frame[DEATHS].values\n",
    "frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Liberia\n",
    "\n",
    "### Check structures of the files\n",
    "\n",
    "#### Headers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'check_headers' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-b490bad5fada>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mallFiles\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mglob\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mglob\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mDATA_FOLDER\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m'liberia_data/*.csv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mcheck_headers\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mallFiles\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'check_headers' is not defined"
     ]
    }
   ],
   "source": [
    "allFiles = glob.glob(DATA_FOLDER + 'liberia_data/*.csv')\n",
    "\n",
    "check_headers(allFiles)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We wanted to use first the \"Newly reported deaths\" column, but they were disrepedancies with the cumulated data, and the newly reported data did not indicate if it was confirmed, suspected, etc. For the \"Total death/s in confirmed cases\",  less than half of the files didn't contain the value for this field. So either we switch to no data or more data but overestimation. Then we decided to switch to cumulated \"Total death/s in confirmed, probable, suspected cases\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[True, True, True, True]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "NEW_CASES = \"New case/s (confirmed)\"\n",
    "TOTAL_NEW_CASES = \"Total confirmed cases\"\n",
    "TOTAL_DEATH = \"Total death/s in confirmed, probable, suspected cases\"\n",
    "NEWLY_REPORTED_DEATH = \"Newly reported deaths\"\n",
    "\n",
    "\n",
    "# most of error are extra \\n and spaces\n",
    "#columns = list(map(lambda x: re.sub(\"\\s\\s+\" , \" \", x.replace(\"\\n\", \"\")), df.index.levels[0].values.tolist()))\n",
    "def check_column_in_files(files, column_name):\n",
    "    for file_ in allFiles:\n",
    "        df = pd.read_csv(file_, index_col=['Variable', 'Date']).fillna(value=0)\n",
    "        \n",
    "        #df[\"edited_variable\"] = df.index.levels[0].values\n",
    "        #df[\"edited_variable\"] = \n",
    "        new_values = list(map(lambda x: re.sub(\"\\s\\s+\" , \" \", x.replace(\"\\n\", \"\")), df.index.levels[0].values.tolist()))\n",
    "        df.index = df.index.set_levels(levels=new_values, level=0)\n",
    "        \n",
    "        #print(df.index.levels[0])\n",
    "        \n",
    "        #result = False\n",
    "        result = column_name in df.index\n",
    "            \n",
    "        \n",
    "        \n",
    "        if(column_name == TOTAL_DEATH):\n",
    "            pass\n",
    "            #print(\"{} {}\".format(result, file_))\n",
    "        if not result:\n",
    "            if(column_name == TOTAL_DEATH):\n",
    "                print(df.index.levels[0])\n",
    "            return False\n",
    "    \n",
    "        \n",
    "    return True\n",
    "\n",
    "\n",
    "to_check = [NEW_CASES, TOTAL_DEATH, NEWLY_REPORTED_DEATH, TOTAL_NEW_CASES]\n",
    "list(map(lambda x: check_column_in_files(allFiles, x), to_check))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, the both lines are available in each file, however a quick check the files indicates that the cell may be empty.\n",
    "\n",
    "Below, we found out that there is many NA values for the lasts months, only totals are available which mixes confirmed, probable and suspected cases. We have chosen along the way to consider only confirmed cases, so we are going to drop those data. First, let's check how many data we are going to drop.\n",
    "\n",
    "Also, there is discrependencies between New deaths reported and cumulated data.\n",
    "\n",
    "For the new cases, we also are going to use the cumulated data, except for the last file, were it seems that the cumulated data has moved to the new cases column. We patch manually this"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the `2014-10-04-v142.csv` file, there is two Total deaths line, we keep only the second one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NaN Death ratio 3.0 % (3/100)\n",
      "NaN New Case ratio 15.0 % (15/100)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total death</th>\n",
       "      <th>New cases</th>\n",
       "      <th>New cases (cumul.)</th>\n",
       "      <th>corr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2011-06-16</th>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-06-29</th>\n",
       "      <td>49.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-07-01</th>\n",
       "      <td>61.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-07-26</th>\n",
       "      <td>145.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>92.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-08-02</th>\n",
       "      <td>257.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-08-28</th>\n",
       "      <td>885.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>369.0</td>\n",
       "      <td>369.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-09-01</th>\n",
       "      <td>1004.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>406.0</td>\n",
       "      <td>406.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-09-30</th>\n",
       "      <td>2058.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>925.0</td>\n",
       "      <td>925.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-10-01</th>\n",
       "      <td>2082.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>926.0</td>\n",
       "      <td>926.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-10-31</th>\n",
       "      <td>2697.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73.0</td>\n",
       "      <td>2437.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-11-02</th>\n",
       "      <td>2700.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2456.0</td>\n",
       "      <td>2456.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-11-30</th>\n",
       "      <td>3157.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2823.0</td>\n",
       "      <td>2823.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-12-01</th>\n",
       "      <td>3161.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2823.0</td>\n",
       "      <td>2823.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2011-12-09</th>\n",
       "      <td>3290.0</td>\n",
       "      <td>2946.0</td>\n",
       "      <td>2946.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Total death  New cases  New cases (cumul.)    corr\n",
       "2011-06-16         16.0        1.0                12.0     NaN\n",
       "2011-06-29         49.0        2.0                43.0    43.0\n",
       "2011-07-01         61.0        2.0                48.0    48.0\n",
       "2011-07-26        145.0        6.0                92.0    92.0\n",
       "2011-08-02        257.0        0.0               132.0   132.0\n",
       "2011-08-28        885.0       22.0               369.0   369.0\n",
       "2011-09-01       1004.0       14.0               406.0   406.0\n",
       "2011-09-30       2058.0        1.0               925.0   925.0\n",
       "2011-10-01       2082.0        0.0               926.0   926.0\n",
       "2011-10-31       2697.0        NaN                73.0  2437.0\n",
       "2011-11-02       2700.0        NaN              2456.0  2456.0\n",
       "2011-11-30       3157.0       10.0              2823.0  2823.0\n",
       "2011-12-01       3161.0        1.0              2823.0  2823.0\n",
       "2011-12-09       3290.0     2946.0              2946.0     NaN"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allFiles = glob.glob(DATA_FOLDER + 'liberia_data/*.csv')\n",
    "\n",
    "list_ = []\n",
    "\n",
    "for file_ in allFiles:\n",
    "    df = pd.read_csv(file_, index_col=['Variable', 'Date'])\n",
    "    \n",
    "    # Normalize variable column\n",
    "    new_values = list(map(lambda x: re.sub(\"\\s\\s+\" , \" \", x.replace(\"\\n\", \"\")), df.index.levels[0].values.tolist()))\n",
    "    df.index = df.index.set_levels(levels=new_values, level=0)\n",
    "    \n",
    "    data = df[[\"National\"]]\n",
    "\n",
    "    \n",
    "    total_deaths = data.loc[TOTAL_DEATH]\n",
    "    new_cases = data.loc[NEW_CASES]\n",
    "    total_cases = data.loc[TOTAL_NEW_CASES]\n",
    "    \n",
    "    concatenated = pd.concat([total_deaths.tail(n=1), new_cases, total_cases], axis=1)\n",
    "    concatenated.columns = [\"Total death\", \"New cases\", \"New cases (cumul.)\"]\n",
    "    \n",
    "    \n",
    "    \n",
    "    list_.append(concatenated)\n",
    "    \n",
    "    \n",
    "    \n",
    "frame = pd.concat(list_)\n",
    "#frame.index = pd.to_datetime(frame.index, format=\"%m/%d/%Y\")\n",
    "\n",
    "#frame.index = frame.index.set_levels([frame.index.levels[0], pd.to_datetime(frame.index.levels[1])])\n",
    "#frame = frame.swaplevel()\n",
    "\n",
    "# keep only NaN value.\n",
    "death_null = frame[frame[\"Total death\"].isnull()].shape[0]\n",
    "death_total = frame.shape[0]\n",
    "\n",
    "cases_null = frame[frame[\"New cases\"].isnull()].shape[0]\n",
    "cases_total = frame.shape[0]\n",
    "print(\"NaN Death ratio {} % ({}/{})\".format(death_null / death_total * 100, death_null, death_total))\n",
    "print(\"NaN New Case ratio {} % ({}/{})\".format(cases_null / cases_total * 100, cases_null, cases_total))\n",
    "\n",
    "frame[\"New cases (cumul.)\"][\"12/9/2014\"] = frame[\"New cases\"][\"12/9/2014\"]\n",
    "#frame[\"New cases (cumul.)\"] = \n",
    "\n",
    "frame.index = pd.to_datetime(frame.index)\n",
    "\n",
    "def remove_outliers(x,y):\n",
    "    #print(x, y, x > y)\n",
    "    if(x > y):\n",
    "        return x\n",
    "    return y\n",
    "\n",
    "frame[\"corr\"] = frame[\"New cases (cumul.)\"].rolling(window=2).apply(lambda x: remove_outliers(x[0], x[1]))\n",
    "\n",
    "frame[\"Month\"] = frame.index.map(lambda x: x.month).astype(int)\n",
    "frame[\"Day\"] = frame.index.map(lambda x: x.day).astype(int)\n",
    "frame.index = [frame[\"Month\"], frame[\"Day\"]]\n",
    "\n",
    "#with pd.option_context('display.max_rows', None, 'display.max_columns', 5):\n",
    "    #print(frame)\n",
    "    \n",
    "a = pd.concat(\n",
    "    [frame.groupby(['Month'])['Day'].min().reset_index(),\n",
    "    frame.groupby(['Month'])['Day'].max().reset_index()]\n",
    ").sort_index()\n",
    "result = pd.merge(a, frame, on=[\"Month\", \"Day\"], how=\"left\")\n",
    "\n",
    "result.index = pd.to_datetime(result[['Month','Day']].apply(lambda x : '{}-{}-2011'.format(x[0],x[1]), axis=1))\n",
    "result = result.drop([\"Month\", \"Day\"], axis=1)\n",
    "\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Write your answer here"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
